pkg/cmd/experimental/syncgroups/config/types.go

```diff
@@ -0,0 +1,204 @@
+
+const (
+	// LDAPSyncTypeFirstClassGroups defines an LDAP schema where groups are stored as first-class LDAP
+	// entries with an attribute containing a list of members, either with members listed with their
+	// DN in the attribute members listed with some other UID in the attribute.
+	LDAPSchemaTypeFirstClassGroups LDAPSchemaType = "FirstClassGroups"
+
+	// LDAPSyncTypeFirstClassUsers defines an LDAP schema where users are stored as first-class LDAP
+	// entries with an attribute containing a list of groups they are a member of, with groups listed
+	// by a unique name, with no additional group metadata in the LDAP server.
+	LDAPSchemaTypeFirstClassUsers LDAPSchemaType = "FirstClassUsers"
+
+	// LDAPSyncTypeFirstClassGroupsAndUsers defines an LDAP schema where users are stored as first-class
+	// LDAP entries with an attribute containing a list of groups they are a member of, either with
+	// groups listed by DN, with the first-class group entries holding additional metadata groups listed
+	// by some other UID, with the first-class group entries holding additional metadata.
+	LDAPSchemaTypeFirstClassGroupsAndUsers LDAPSchemaType = "FirstClassGroupsAndUsers"
+)
+
+// GroupSyncScope determines the scope of the group sync operation
+type GroupSyncScope string
+
+const (
+	// GroupSyncScopeAll determines that all groups from the source should be synced
+	GroupSyncScopeAll GroupSyncScope = "All"
+	// GroupSyncScopeWhitelist determines that a whitelist of groups from the source should be synced
+	GroupSyncScopeWhitelist GroupSyncScope = "Whitelist"
+)
+
+// GroupSyncSource determines the source of the groups to be synced
+type GroupSyncSource string
+
+const (
+	// GroupSyncSourceLDAP determines that the groups to be synced are determined from an LDAP record
+	GroupSyncSourceLDAP GroupSyncSource = "LDAP"
+	// GroupSyncSourceOpenShift determines that the groups to be synced are determined from OpenShift records
+	GroupSyncSourceOpenShift GroupSyncSource = "OpenShift"
+)
+
+// LDAPSyncConfig contains all of the configuration information for an LDAP group sync job.
+type LDAPSyncConfig struct {
+	// ServerConfig holds LDAP server information
+	ServerConfig ldaputil.LDAPServerConfig
+
+	// GroupClient allows for updates to etcd records
+	GroupClient client.GroupInterface // added to loaded config upon cli invocation
+
+	// Source determines the source of the list of groups to sync
+	Source GroupSyncSource // added to loaded config from cli input
+
+	// Scope determines the scope of the group sync
+	Scope GroupSyncScope // added to loaded config from cli input
+
+	// WhitelistContents are the contents of the whitelist - either names of OpenShift groups given by
+	// an administrator on the command line invocation of the sync-groups command or LDAP group UIDs
+	// read from a whitelist file
+	WhitelistContents []string // added to loaded config from cli input
+
+	// Type determines which LDAP server schema exists for the sync
+	Type LDAPSchemaType
```

Comment:
This is something I borrowed from our `buildapi` - it was a very common thing there. Pros/cons? Clearly having `<nil>` parts of the config object isn't the best, but otherwise we will have three objects that are exactly the same save this one part.